### PR TITLE
Remove transmute from `btree::node::Node::as_slices_internal_mut`

### LINKED ISSUE #27620: btree: Reliance on structure layout
The function `collections::btree::node::Node::as_slices_internal_mut` relies on structure layout of two relatively unrelated structures to be the same, which is considered a bad thing™:

    #[inline]
    pub fn as_slices_internal_mut<'b>(&'b mut self) -> MutNodeSlice<'b, K, V> {
        unsafe { mem::transmute(self.as_slices_internal()) }
    }

### PR DESCRIPTION
**Pull request by apasel422**
Closes #27620.

### CODE REVIEW

File: src/libcollections/btree/node.rs [hunk 1]
```diff
@@ -357,7 +357,10 @@ impl<K, V> Node<K, V> {
 
     #[inline]
     pub fn as_slices_mut<'a>(&'a mut self) -> (&'a mut [K], &'a mut [V]) {
-        unsafe { mem::transmute(self.as_slices()) }
+        unsafe {(
+            slice::from_raw_parts_mut(*self.keys, self.len()),
+            slice::from_raw_parts_mut(*self.vals, self.len()),
```
Comment (nagisa) [inline src/libcollections/btree/node.rs]: I think this might be losing aliasing information.

Comment (apasel422) [inline src/libcollections/btree/node.rs]: Any suggestions?

**Comment by alexcrichton** [comment]
cc https://github.com/rust-lang/rust/pull/27252, there's a bit of duplication between these two

**Comment by apasel422** [comment]
Rebased.

**Comment by apasel422** [comment]
Does this need any further changes?

**Comment by Gankro** [comment]
Oh whoops, I thought this has been accepted.

@bors r+

**Pull request merged**